### PR TITLE
reports: add SARIF reporting

### PR DESCRIPTION
**Pull request by de-jcup**
- initial SARIF report implementation 
  ( all parts of the implementation are in own folders/packages, no changes on other reporting parts)
- added tests
- automated shrinking of web response and request body when necessary 
  (currently 8k webrequest or 32k webresponse are allowed before shrinking)
- automated binary/text detection
- hiding authorization header information in SARIF Report with asterisks
- using MITRE CWE 4.8 taxonomy

Fix zaproxy/zaproxy#6875

### CODE REVIEW

**Review by psiinon — CHANGES_REQUESTED** [review]
It generates valid JSON, not sure if its valid SARIF or not 😉 
A couple of changes are required - the unit tests dont run for me because they just end in "Test" - they should end in "UnitTest".
Also the help needs to be updated: https://github.com/zaproxy/zap-extensions/blob/main/addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/templates.html - it should include an example fragment of the report.

**Comment by de-jcup** [comment]
I added currently only the minimum help files (`templates.html` + `reporting-sarif-json.html`) and updated the change log.

Had some issues with the java help files:
- It is clear that we are talking about the "java help " (https://github.com/javaee/javahelp/) 
- But while inspecting some of the different help i18n folders i saw that they just contain the exact content like the default help folder (e.g. `help_de_DE` does contain nothing in german). It is a little bit irritating that there is so much duplicated content. Does a developer need to copy all the stuff into the folders manually - or is there a special gradle task available for this?
- Looking at https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/ I found no additional info about the help structure/ special build commands for the help files.

**Comment by thc202** [comment]
https://github.com/zaproxy/zap-extensions/blob/main/CONTRIBUTING.md#help

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java
Comment (thc202) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:26]: Remove these, this does not work.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java
Comment (thc202) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:33]: `Zed`

**Comment by psiinon** [comment]
FYI the help is now good 😁

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:26]: @thc202 : Please explain in more detail what you want to have changed/what exactly is not working.

Comment (thc202) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:26]: What psiinon said, remove those comments, they don't work, formatting happens always.

Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:26]: Ah. Thank you. In our project we use the eclipse formatter together with spotless - here the formatter:on /off flags do work. Will remove all of those comments in code.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java
Comment (thc202) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:44]: Was the last slash added on purpose?

File: addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json
Comment (thc202) [inline addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:97]: Same here.

File: addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json
Comment (thc202) [inline addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:121]: Same here.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java
Comment (thc202) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java]: Not all of them were removed yet e.g. `SarifReportDataSupportUnitTest`.

File: addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json
Comment (kingthorin) [reply addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:94]: Is there some sort of quoted string builder/method that should be used?

Comment (thc202) [reply addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:94]: https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html#inlining use of `[[` vs `[(`.

Comment (thc202) [reply addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:94]: Ideally tests should catch this.

Comment (kingthorin) [reply addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:94]: Perfect, I just wasn't aware because this isn't something I've worked on yet.

File: addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json
Comment (de-jcup) [reply addOns/reports/src/main/zapHomeFiles/reports/sarif-json/report.json:121]: I will extend the report integration test with problematic chars inside. Will also try to change all occurrences of "[(" to "[[" where possible.

**Comment by de-jcup** [comment]
@psiinon :
> It generates valid JSON, not sure if its valid SARIF or not wink 

https://sarifweb.azurewebsites.net/Validation is a way to check easily if SARIF is valid or not.
It also explains problems and has a link to the corresponding OASIS documentation.

Because the "Dev Build" ZAP version will result currently in non semantic version, I will improve the report here as well, so that reports created by dev builds have a semantic version (fallback to "0.0.0")

**Comment by de-jcup** [comment]
https://sarifweb.azurewebsites.net/Validation does accept the SARIF reports from ZAP without any validation failures.

- Via the online validator It's now relative easy to check if the output report is valid SARIF
- In case of "Dev Build" Zap versions, the report does now contain "0.0.0-Dev Build" as version and also as "semantic version". For real versions, the origin ZAP version will be used
- When start line is invalid (not >=1) in a text region (because of snippet parse problems) there will always be a fallback to 1 and and an additional property "startLineFailure" inside the region with a detailed desription in this case. This also ensures, that even in such an odd situation (should not happen) the SARIF output is still valid.
- Also following bugfix: the taxonomy version is now no longer a static value (4.4) but the real one

**Review by psiinon — APPROVED** [review]
LGTM

File: addOns/reports/CHANGELOG.md
Comment (kingthorin) [inline addOns/reports/CHANGELOG.md:14]: This should be moved to the `Unreleased` section.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: Shouldn't these use `{@code null}`?

Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: Throughout.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:63]: Should the leading `/` be removed to allow more types?

Ex those that are +xml or +json: https://github.com/zaproxy/zaproxy/blob/45180ea383ec85317d3f342f2c86763b337a8172/docker/scripts/scripts/httpsender/Alert_on_Unexpected_Content_Types.js#L15-L30

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:54]: Should this include a yaml check?

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:52]: -> `isJson`

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:53]: -> `isXml`

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:39]: Note: commonlib does have a method for checking whether or not responses contain control chars, which may be of benefit here? (If the actual content, not just the header, is available.)

https://github.com/zaproxy/zap-extensions/blob/d7b0b875a37f40cbf3d6c95d9054e5e232128614/addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/ResourceIdentificationUtils.java#L112-L117

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifGuid.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifGuid.java:45]:
```suggestion
    public static SarifGuid createToolcomponentGuid(SarifTaxonomyDataProvider component) {
```

Should the "c" on component be uppercase?

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifGuid.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifGuid.java:59]:
```suggestion
     * Creates a SARIF guid object by using data from taxonomy provider and the given taxonomy id.
```

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java:77]: I believe this should use 
```suggestion
        return name().toLowerCase(Locale.ROOT);
```

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifMessage.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifMessage.java:29]: -> `setContentAsHtml`

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java:23]:
```suggestion
 * Internal class to hold protocol and version separated. Also provides parse functionality.
```

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java:32]: Per the comment shouldn't the variables and methods be static?

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifReportDataSupport.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifReportDataSupport.java:50]:
```suggestion
    // same report will produce same ordering etc.
```

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifReportDataSupport.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifReportDataSupport.java:114]: -> `createCweTaxonomy`

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifRule.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifRule.java:52]: Use `String.valueOf` instead of weird concat?

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifRule.java
Comment (kingthorin) [inline addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifRule.java:156]: i18n?

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetectorUnitTest.java
Comment (kingthorin) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetectorUnitTest.java:77]: Add some that ensure charset is handled or not mis-parsed?
`Content-Type: text/html; charset=utf-8`

File: addOns/reports/CHANGELOG.md
Comment (de-jcup) [reply addOns/reports/CHANGELOG.md:14]: Hmm...
I think there is also a bug inside the markdown indention of the existing change log:


> # Changelog
> All notable changes to this add-on will be documented in this file.
> 
> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
> 
> ## Unreleased
> # Changed
> - The XML and JSON reports now include programName metadata elements (Issue 6640).
> 
> ## [0.16.0] - 2022-10-27
> ### Added
> 
According to https://keepachangelog.com/en/1.0.0/ I will change it to
```
## Unreleased
# Changed
```
to 

```
## Unreleased
### Added
- SARIF reporting

### Changed
```
And also add a `###Added`below `##Unreleased`

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: I am not sure what here is meant. You marked javadoc and mentioned `{@code null}` - please clarify.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:54]: According to the official MEME type list from IANA at https://www.iana.org/assignments/media-types/media-types.xhtml there is no official yaml MIME type. 
Have you seen something like "text/yaml" in the real live? We can add it, but AFAIK you have to use "text/plain" for yaml files.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java
Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: It should use a javadoc tag not html

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:63]: I was regarding to https://www.iana.org/assignments/media-types/media-types.xhtml - and there is always a leading slash before the type.

I tried to catch all official types in a correct way - is there something used in the wild which would not be catched by the implementation?

File: addOns/reports/CHANGELOG.md
Comment (kingthorin) [reply addOns/reports/CHANGELOG.md:14]: Yup the headings are currently wrong

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:54]: I hadn’t looked it up, I just assumed there was one. Do t worry about it I guess.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:63]: Yes, there are examples in the linked code of things encountered in the wild.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java:77]: We do not need this here, because it's just the name of the `SarifLevel` enumeration and the names are always locale neutral.

Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java:77]: Ok

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java:32]: I will change the comment to
```
 // private constructor - so static parse method must be used to create the protocol data object
```
so it is more clear: You have different objects, so getters may not be static, just the creation is done in a static method, so by the name of the method it is clear how the protocol data is being constructed.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java
Comment (thc202) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java:77]: The comment is contradictory (or I'm not understanding it correctly), if you say it's neutral it should use a neutral `Locale`, otherwise it's going to use the default (e.g. in Turkish locale `WARNING` will be lower cased to `warnıng`, note the `ı`).

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java
Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java:32]: Ok

Comment (thc202) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifProtocolData.java:32]: Or just remove the comment, that the constructor is private already indicates that other method needs to be used.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifRule.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifRule.java:156]: SARIF has no official confidence element but there is a general `properties` element which can be used by vendors/products to add additional content. 

So I decided to add `confidence` to the report by using `properties` element.

About the I18n: I do not want to add i18n here because
- Those values shall be readable by a machine
- We would need to add something like a "confidence-language" property as well to describe which language is used  and maybe something like a  "confidence-i18n" propertty in addition to have it still machine readable by another product (in our case SecHub)
- Properties are just a key,value map and there is no easy possibility to describe a structure (to avoid the former mentioned additional property keys)
- Avoid effort in transalations
- Keep it simple

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: I use `<code>` a lot in javadoc and was not aware about that this shall not be used in ZAP.
When looking inside the existing ZAP source code I found 195 occurrences of "<code>" (including my ones) inside the workspace.

So shall I change all of my `<code>` elements? And what about `<h3>` and others?

Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: I hadn’t looked around. I’ll defer to others. @thc202 do you have preference?

Comment (thc202) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: Use `{@code }` less verbose, it's not about the HTML generally.

Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:38]: Found 7 matches in my code - will replace all parts with the `{@code null}` variant.

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java:78]: @thc202 : Thanx a lot! 

I was absolutely nota ware, that standard ASCII charactors (0-127) could be locale specific when it comes to a simple `toLowercase()` call!
I tried it out with TR locale and you are right: It really does change it - I will definitely change this to the version @kingthorin suggested. I will look at other locations in my code as well and fix this as well (if there are any.

Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifLevel.java:78]: Thanks

File: addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java
Comment (de-jcup) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:39]: Hm.. interesting, but I think -  at least as a first version of the SARIF report handling - we should rely on the content type returned by the server. I am not sure if the existence of an ISO control character is really the 100% way to accept the content as a binary or not. To avoid edge cases, I would keep it as is.

Comment (kingthorin) [reply addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:45]: That’s fine

**Comment by joubin** [comment]
@psiinon @kingthorin is there anything specific pending before we can merge this?

**Review by kingthorin — APPROVED** [review]

File: addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-sarif-json.html
Comment (thc202) [inline addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-sarif-json.html:4]: Needs to be updated.

Comment (de-jcup) [reply addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-sarif-json.html]: @thc202 : I did change this in https://github.com/zaproxy/zap-extensions/pull/4078/commits/eefc4d7610b8653301dc8cade47646b5977af458

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**
